include/xtensor/xtiny.hpp

```diff
@@ -0,0 +1,1925 @@
+            {
+                // self uses the buffer, other allocated memory
+                for(size_type k=0; k<m_size; ++k)
+                {
+                    other.m_buffer[k] = m_data[k];
+                }
+                m_data = other.m_data;
+                other.m_data = other.m_buffer;
+            }
+            else
+            {
+                // both use the buffer
+                if(m_size < other.m_size)
+                {
+                    for(size_type k=0; k<m_size; ++k)
+                    {
+                        swap(m_data[k], other.m_data[k]);
+                    }
+                    for(size_type k=m_size; k<other.m_size; ++k)
+                    {
+                        m_data[k] = other.m_data[k];
+                    }
+                }
+                else
+                {
+                    for(size_type k=0; k<other.m_size; ++k)
+                    {
+                        swap(m_data[k], other.m_data[k]);
+                    }
+                    for(size_type k=other.m_size; k<m_size; ++k)
+                    {
+                        other.m_data[k] = m_data[k];
+                    }
+                }
+            }
+        }
+        swap(m_size, other.m_size);
+    }
+
+    template <class V, index_t B>
+    inline void
+    xtiny_impl<V, runtime_size, V[B]>::allocate(value_type const & v)
+    {
+        if(m_size > buffer_size)
+        {
+            m_data = m_allocator.allocate(m_size);
+            std::uninitialized_fill(m_data, m_data+m_size, v);
+        }
+        else
+        {
+            std::fill(m_data, m_data+m_size, v);
+        }
+    }
+
+    template <class V, index_t B>
+    inline void
+    xtiny_impl<V, runtime_size, V[B]>::allocate(tags::skip_initialization_tag)
+    {
+        if(m_size > buffer_size)
+        {
+            m_data = m_allocator.allocate(m_size);
+            if(!may_use_uninitialized_memory)
+            {
+                std::uninitialized_fill(m_data, m_data+m_size, value_type());
+            }
+        }
+    }
+
+    template <class V, index_t B>
+    template <class IT, class>
+    inline void
+    xtiny_impl<V, runtime_size, V[B]>::allocate(IT begin)
+    {
+        if(m_size > buffer_size)
+        {
+            m_data = m_allocator.allocate(m_size);
+            for(size_type k=0; k<m_size; ++k, ++begin)
+            {
+                m_allocator.construct(m_data+k, static_cast<value_type>(*begin));
+            }
+        }
+        else
+        {
+            for(size_type k=0; k<m_size; ++k, ++begin)
+            {
+                m_data[k] = static_cast<value_type>(*begin);
+            }
+        }
+    }
+
+    template <class V, index_t B>
+    inline void
+    xtiny_impl<V, runtime_size, V[B]>::deallocate()
+    {
+        if(m_size > buffer_size)
+        {
+            if(!may_use_uninitialized_memory)
+            {
+                for(size_type k=0; k<m_size; ++k)
```

Comment:
maybe we could reuse `detail::safe_destroy_deallocate` from xstorage here?

Reply:
I considered this, but then found that I'd first like to change `safe_init_allocate()`:
```
// rename current version
// (*_init_* is a misnomer, because the function doesn't init trivial types)
safe_allocate_no_init(A& alloc, typename A::size_type size)

// add new version with custom initial value
safe_allocate_init(A& alloc, typename A::size_type size, 
                   typename A::value_type const & v =  typename A::value_type())
```
Then I could use `safe_allocate_*` and `safe_destroy_deallocate` symmetrically.

Reply:
I looked up where `safe_init_allocate` is called in the current code base, and found this:
```c++
pointer tmp = safe_init_allocate(al, rhs.m_size);
if (xtrivially_default_constructible<value_type>::value)
{
    std::uninitialized_copy(rhs.m_data.get(), rhs.m_data.get() + rhs.m_size, tmp);
}
else
{
    std::copy(rhs.m_data.get(), rhs.m_data.get() + rhs.m_size, tmp);
}
```
This code is awkward, because `std::uninitialized_copy` works for arbitrary types, not just trivial ones. The above could simply be replaced with:
```c++
pointer tmp = al.allocate(rhs.m_size);
std::uninitialized_copy(rhs.m_data.get(), rhs.m_data.get() + rhs.m_size, tmp);
```

Reply:
I didn't write that code, but maybe @JohanMabille has some insights here?

Reply:
`uninitialized_copy` calls the constructor, which should have already been called by `safe_init_allocate` when the type is not default constructible (hence the `safe` inthe name). So using `uninitialized_copy` everywhere may end with a double call to the constructor with all the bad consequences it can have.

Reply:
I don't suggest to use `uninitialized_copy` everywhere. But in the particular code snippet above, non-trivial objects are needlessly initialized twice, first with placement new, then with copy assignment. In contrast, `al.allocate(rhs.m_size);` followed by `uninitialized_copy` always calls placement new exactly once, as it should be.

Reply:
Sorry for the word "everywhere", I meant in both branches of the "if" of the previous code snippet. I agree that in that case the code can be simplified. We loose the "symmetric" appearance of the code (with the use of safe_destroy_deallocate), but the code is cleaner.

However, I don't understand why you would like to have two functions; you would have to test if the type is trivially default constructible to know which method to call, while the idea in safe_init_allocate is to automatically handle that and default construct the values only if it is needed. The "safe" here is related to the initialization, not the allocation, and it is safe to no initialize trivial types.

Reply:
> it is safe to not initialize trivial types.

Exactly that's why I thought that a name containing "no" (`safe_no_init_allocate` or `safe_allocate_no_init` or `allocate_safe_no_init`) would more explicitly express that trivial types are safely *not* initialized 😄 .

> why you would like to have two functions

When I free memory with `safe_destroy_deallocate` (which I would like to do), I also want to allocate it with a corresponding allocate function (general rule: always allocate and free memory with corresponding function pairs). Since this must apply to all supported `xtiny` constructors, I actually need three allocate variants, one for safe (non-)initialization, value initialization, and copy initialization. In the hypothetical case that someone modifies the allocation/deallocation mechanism in `xstorage`, all functions in that file will hopefully be changed together, whereas a subsequent change in a seemingly unrelated file like `xtiny` is easily overlooked. Alternatively, `xstorage` could explicitly document a guarantee that incompatible changes of the allocation/deallocation mechanism will never occur.